apps/wpcom-block-editor/src/wpcom/features/tracking.js

```diff
@@ -110,16 +110,42 @@ const getBlocksTracker = ( eventName ) => ( blockIds ) => {
 	} );
 };
 
+/**
+ * Determines whether a block pattern has been inserted and if so, records
+ * a track event for it. The recorded event will also reflect whether the
+ * inserted pattern replaced blocks.
+ *
+ * @param {Array} args Data supplied to block insertion or replacement tracking functions.
+ * @returns {string} Pattern name being inserted if available.
+ */
+const trackPatternInsertion = ( args ) => {
+	const meta = find( args, ( arg ) => arg?.patternName );
+	const patternName = meta?.patternName;
+
+	if ( patternName ) {
+		tracksRecordEvent( 'wpcom_pattern_inserted', {
+			pattern_name: patternName,
+			blocks_replaced: args.blocks_replaced,
+		} );
+	}
+
+	return patternName;
+};
+
 /**
  * Track block insertion.
  *
  * @param {object|Array} blocks block instance object or an array of such objects
+ * @param {Array} args additional insertBlocks data e.g. metadata containing pattern name.
  * @returns {void}
  */
-const trackBlockInsertion = ( blocks ) => {
+const trackBlockInsertion = ( blocks, ...args ) => {
+	const patternName = trackPatternInsertion( { ...args, blocks_replaced: false } );
```

Comment:
Minor but when I read through this I thought `trackPatternInsertion` was always tracking when in fact it's optional.

Maybe rename to `maybeTrackPatternInsertion`? Not a big deal but just a thought...